File: frontend/src/app/api/maintenance.ts
import { apiRawRequest } from "./client";

export type MaintenanceOrderRow = {
  id: number;
  order_number?: string | null;
  customer_name: string;
  customer_phone?: string | null;
  device_type: string;
  device_brand?: string | null;
  device_model?: string | null;
  serial_number?: string | null;
  problem_description: string;
  diagnosis?: string | null;
  status: string;
  priority: string;
  estimated_cost?: number;
  final_cost?: number;
  paid_amount?: number;
  technician?: string | null;
  warranty_days?: number;
  notes?: string | null;
  received_at?: string;
  completed_at?: string | null;
  delivered_at?: string | null;
};

export type MaintenanceCreatePayload = {
  customer_name: string;
  customer_phone?: string;
  device_type: string;
  device_brand?: string;
  device_model?: string;
  serial_number?: string;
  problem_description: string;
  priority?: string;
  estimated_cost?: number;
  notes?: string;
  branch_id?: number;
};

export type MaintenanceUpdatePayload = {
  diagnosis?: string;
  status?: string;
  final_cost?: number;
  paid_amount?: number;
  technician?: string;
  notes?: string;
};

export type MaintenanceHistoryRow = {
  id: number;
  order_id: number;
  old_status?: string | null;
  new_status?: string | null;
  notes?: string | null;
  changed_by?: string | null;
  changed_at: string;
};

export function getMaintenanceOrders(token: string, status?: string) {
  const query = status ? `?status=${encodeURIComponent(status)}` : "";
  return apiRawRequest<MaintenanceOrderRow[]>(`/maintenance/${query}`, { token });
}

export function createMaintenanceOrder(token: string, payload: MaintenanceCreatePayload) {
  return apiRawRequest<{ id: number }>("/maintenance/", {
    method: "POST",
    token,
    body: payload
  });
}

export function getMaintenanceOrder(token: string, id: number) {
  return apiRawRequest<MaintenanceOrderRow>(`/maintenance/${id}`, { token });
}

export function updateMaintenanceOrder(token: string, id: number, payload: MaintenanceUpdatePayload) {
  return apiRawRequest<{ message: string }>(`/maintenance/${id}`, {
    method: "PUT",
    token,
    body: payload
  });
}

export function getMaintenanceHistory(token: string, id: number) {
  return apiRawRequest<MaintenanceHistoryRow[]>(`/maintenance/${id}/history`, { token });
}

File: app/models/maintenance.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List

class MaintenanceOrderBase(BaseModel):
    order_number: Optional[str] = None
    customer_name: str
    customer_phone: Optional[str] = None
    device_type: str
    device_brand: Optional[str] = None
    device_model: Optional[str] = None
    serial_number: Optional[str] = None
    problem_description: str
    diagnosis: Optional[str] = None
    status: str = "received"
    priority: str = "normal"
    estimated_cost: float = 0
    final_cost: float = 0
    paid_amount: float = 0
    parts_used_json: str = "[]"
    technician: Optional[str] = None
    warranty_days: int = 0
    notes: Optional[str] = None
    branch_id: int = 1

class MaintenanceOrderCreate(MaintenanceOrderBase):
    pass

class MaintenanceOrderUpdate(BaseModel):
    diagnosis: Optional[str] = None
    status: Optional[str] = None
    final_cost: Optional[float] = None
    paid_amount: Optional[float] = None
    parts_used_json: Optional[str] = None
    technician: Optional[str] = None
    completed_at: Optional[str] = None
    delivered_at: Optional[str] = None
    notes: Optional[str] = None

class MaintenanceOrder(MaintenanceOrderBase):
    id: int
    received_at: str
    completed_at: Optional[str] = None
    delivered_at: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

# ── Extension Models ──

class MaintenanceStatusLog(BaseModel):
    id: int
    order_id: int
    old_status: Optional[str]
    new_status: Optional[str]
    notes: Optional[str]
    changed_by: Optional[str]
    changed_at: str

    model_config = ConfigDict(from_attributes=True)

class MaintenanceImage(BaseModel):
    id: int
    order_id: int
    kind: str
    original_name: Optional[str]
    stored_path: str
    created_at: str

    model_config = ConfigDict(from_attributes=True)

class DeviceModelCatalog(BaseModel):
    id: Optional[int] = None
    device_type: str
    model: str
    brand: Optional[str] = ""
    notes: Optional[str] = None
    active: int = 1

    model_config = ConfigDict(from_attributes=True)

class MaintenanceInvoiceVersion(BaseModel):
    id: int
    order_id: int
    version: int
    reason: Optional[str]
    payload_json: str
    created_by: Optional[str]
    created_at: str

    model_config = ConfigDict(from_attributes=True)
